src/main/kotlin/cacophony/regex/RegexUtil.kt

```diff
@@ -0,0 +1,49 @@
+package cacophony.regex
+
+import cacophony.utils.AlgebraicRegex
+import cacophony.utils.AlgebraicRegex.AtomicRegex
+import cacophony.utils.AlgebraicRegex.ConcatenationRegex
+import cacophony.utils.AlgebraicRegex.StarRegex
+import cacophony.utils.AlgebraicRegex.UnionRegex
+
+class RegexSyntaxErrorException : Exception()
+
+internal sealed class RegexType {
+    abstract fun toAlgebraicRegex(): AlgebraicRegex
+}
```

Comment:
Is it needed? Can't we use already existing Regex classes? If the interface is wrong and conversion is hard, it can be changed.

Reply:
Union and Concat in AlgebraicRegex declare its internals as valargs, therefore does not allow for cheap addition to them (which is used as simple optimalization to reduce nesting of resulting AlgebraicRegex) - as they should btw, AlgebraicRegex should be somehow immutable object in my opiniom.

Other thing is, RegexType is thought as internal to the parser - in theory there may be new operators (eg. ? or []) which AlgebraicRegex does not need, or additional info (eg. position in starting stringi for error printing, as below) needed only for parser.

Think of RegexType as Builder type for AlgebraicRegex